### ISSUE TITLE
UnifiedMap VTM: Remove pseudo-3D display of buildings (or make it configurable)

### ISSUE THREAD

**Issue by moving-bits** [cgeo/cgeo#15723]
Personally I find map display of buildings with UnifiedMap VTM confusing and harder to process than with old Mapsforge. I found out that there's a way to configure 2D/3D, so I propose to switch back to 2D (or make it configurable, if someone prefers 3D):

|3D (currently)|2D (proposed)|
|---|---|
|![image](https://github.com/cgeo/cgeo/assets/3754370/a253786e-f6ee-4e11-bb02-f9d22085bc61)|![image](https://github.com/cgeo/cgeo/assets/3754370/5bc934b8-6ef0-49f8-9bc2-a7b58d51f6c4)|

(screenshots taken with OpenAndroMaps' Berlin map + Elevate theme)

**Comment by eddiemuc** [comment]
I strongly prefer the 3D look of buildings. Please don't remove it completely. Making it configurable would be fine for me though.

**Comment by MagpieFourtyTwo** [comment]
Don't like the 3D effect, too, cause buildings are way harder to recognize.

Which can perhaps be solved by drawing the roofs with a darker shading (already made some attempts to achieve this, but no success yet), but taking into account that there's no real information gain plus most probably higher CPU/GPU load for rendering, I would vote for 2D only, too, although an extra setting would of course be fine, too, but if so, I would suggest to set the default to false.

**Comment by eddiemuc** [comment]
Cruiser has such a setting:
![Screenshot_2024-05-12-11-18-19-21_beed72cb86d752c6022398a83c94aee3.jpg](https://github.com/cgeo/cgeo/assets/6909759/0f3a464f-e1d7-4149-8798-91278ed1af1d)

**Comment by MagpieFourtyTwo** [comment]
Clever wording - cause "normal" is subject to evolutionary shifting. ;)

Nevertheless, just in case: I would prefer something more concrete like "3D buildings" or perhaps even "3D effects". :)

**Comment by devemux86** [comment]
3D buildings can be rendered normally or with shadows and extra OSM tags: roof, color, material, etc.

All these are included in the Mapsforge maps. 3D buildings are declared in the VTM themes.
Mapsforge themes do not declare them, so there is an extra rule in the code with default colors.

![desktop](https://user-images.githubusercontent.com/3484020/210542646-0e5a6bfb-709e-4410-9329-3680068c44cf.png)
![3d_shadows](https://user-images.githubusercontent.com/3484020/53115039-d4c5d680-354d-11e9-951c-8f4202293704.png)

**Comment by eddiemuc** [comment]
Screenshot from c:geo VTM using a Mapsforge Offline map:
![Screenshot_2024-05-12-13-08-27-62_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/cgeo/cgeo/assets/6909759/f3f0b454-488f-4fd0-94f4-8e0e7b370905)

I like it!

**Comment by bekuno** [comment]
> Screenshot from c:geo VTM using a Mapsforge Offline map: !
> ...
> I like it!

Great!
We could integrate the configuration in the menu "Design options".

**Comment by moving-bits** [comment]
PR #15731 adds such an option to VTM offline maps' theme options.

Additional formatting for 3D mode (as shown in screenshot above) should go into a separate PR (and maybe differentiate between 2D / 3D (as today) / 3D (extended, as shown above).

**Comment by moving-bits** [comment]
Basic 2D/3D toggle has been implemented meanwhile, therefore removing my assignment.
If someone wants to implement further 3D options, please go ahead, therefore leaving this issue open.

**Comment by SammysHP** [comment]
@moving-bits I had to turn the option on and off once to disable 3D buildings.

**Comment by moving-bits** [comment]
> @moving-bits I had to turn the option on and off once to disable 3D buildings.

Thanks for reporting. PR #15772 fixes the missing initialization.

**Comment by MagpieFourtyTwo** [comment]
Have switched it off and on several times, even restarted c:geo, but it shows no effect at all, buildings are still pseudo-3D, i. e. hardly visible (most fortunately 3D just kicks in at zoom levels 20 and 10 meters; Map: Offline Openandromaps with Elevate theme).

**Comment by moving-bits** [comment]
Can't reproduce on my emulated device (Pixel 3a, API 30), using OAM Berlin.map with Elevate theme:

|before|setting|after
|---|---|---|
|![Screenshot_1719739561](https://github.com/cgeo/cgeo/assets/3754370/8c5c3fe8-b177-4370-86f6-8c16e5a9a7c5)|![Screenshot_1719739546](https://github.com/cgeo/cgeo/assets/3754370/593675ca-954c-4495-8d12-0d9a4295f680)|![Screenshot_1719739536](https://github.com/cgeo/cgeo/assets/3754370/28370378-60d0-4de9-8f12-e452d49e8c13)|

**Comment by MagpieFourtyTwo** [comment]
This is pretty interesting, because over here it's kind of just the other way round: I get no 3D with zoom levels > 20 m (i. e. 50, 100 ... m), and I always get 3D < 20 m (i. e. the "second" 20 m zoom, 10 and 5 meters). And in both cases the setting doesn't change anything. 

Can provide some screenshots this evening, if needed.

**Comment by eddiemuc** [comment]
> Can't reproduce on my emulated device (Pixel 3a, API 30), using OAM Berlin.map with Elevate theme:

VTM is currently displayed with an orange triangle in the upper left corner. Why is this not in your screenshots? Is this really VTM?

**Comment by moving-bits** [comment]
> > Can't reproduce on my emulated device (Pixel 3a, API 30), using OAM Berlin.map with Elevate theme:
> 
> VTM is currently displayed with an orange triangle in the upper left corner. Why is this not in your screenshots? Is this really VTM?

Yes, definitively VTM, I can rotate OSM map here (and "classic" Mapsforge does have neither the 3D "problem" nor the setting to switch 3D off). I have "hide toolbar on tap" enabled, that's why the orange triangle is hidden behind the toolbar.

**Comment by moving-bits** [comment]
Here are some more screenshots in zoom level "50m" - still 3D over here:

|3D buildings enabled|3D buildings disabled|
|---|---|
|![Screenshot_1719744226](https://github.com/cgeo/cgeo/assets/3754370/362778ce-643f-4ab5-ad59-4e9ddbf6c736)|![Screenshot_1719744388](https://github.com/cgeo/cgeo/assets/3754370/60e3db6b-41c5-4336-a5eb-afe31d4ea219)|


a bit less misleading than in more zoomed-in states, but still 3D

**Comment by MagpieFourtyTwo** [comment]
Apart from that the switch has very limited effects, it additionally seems to depend on chosen Map.
So it behaves different if ``OpenAndroMaps Baden-Württemberg`` or ``Combined offline`` is selected.

Just have a look at the video below, there you will first see Map ``Combined offline`` with ``Show buildings in 3D`` deactivated:

1. Zoom 100 m: Buildings in 2D (and visible backgrounds, BTW)
2. Zoom 50 m: Buildings switch to 3D (although 3D is deactivated)
3. Zoom 20 m; Buildings still 3D
4. Zoom 20 m (second time): Buildings still 3D  (and background vanishes)
5. Zooming out to 100 again: Buildings are 2D now again

Then I switched Map to ``OpenAndroMaps Baden-Württemberg``:

6. Zoom changes to 200 m (for no obvious reason), Buildings are still 2D
7. Zoom to 100 and then 50 m: Buildings still 2D
8. Zoom to 20 m two times: Buildings still 2D (as expected, and Backgrounds still visible)
9. switching Map to ``Combined offline``: Buildings switch to 3D (and Backgrounds vanish)

https://github.com/cgeo/cgeo/assets/9889015/e2adcf54-7afa-4b00-b9ff-09129e800b6f

Next one starting again with Map ``Combined offline``, but now ``Show buildings in 3D`` is activated:
... while the result is in fact nearly the same like above, only difference is, that with ``OpenAndroMaps Baden-Württemberg`` Buildings are now 3D for real - but even here only if Zoom level is <=50 m, >50 it's still 2D:

https://github.com/cgeo/cgeo/assets/9889015/ce3f2e47-a114-4db0-8feb-d8aeea79c871

But, TMALSS: Only thing I want is, to be able to deactivate this so called 3D stuff completely. Cause as can easily be seen in the video, it's pretty hard - and with sight angles of 90° even completely impossible - to distinguish these pseudo-3D-Buildings from the background. So, although this is of course a nice gimmick, it comes at the cost of a massive loss of information, which doesn't really make sense to me.

**Comment by moving-bits** [comment]
@devemux86 
Do you have any idea what could be causing this behavior?

We have a `BuildingLayer` for offline maps, which we enable/disable on applying map theme, according to our "show buildings in 3D" setting. Anything we are missing here?

**Comment by devemux86** [comment]
@moving-bits 

It is indeed the `BuildingLayer` that is added or removed and does the 3D buildings rendering.

It works with all offline maps and VTM or Mapsforge themes.
Note: Mapsforge themes do not contain the extrusion rule, VTM adds it automatically in code.

**Comment by moving-bits** [comment]
> It is indeed the `BuildingLayer` that is added or removed and does the 3D buildings rendering.

for simplification we always add the `BuildingLayer` in VTM, but either enable or disable it, depending on the "show buildings in 3D" setting. - May that be the cause or should it be fine to just enable/disable the layer?

**Comment by devemux86** [comment]
> enable or disable it

It also works like this in [VTM samples](https://github.com/mapsforge/vtm/blob/master/vtm-gdx/src/org/oscim/gdx/InputHandler.java#L280).

**Comment by moving-bits** [comment]
> It also works like this in [VTM samples](https://github.com/mapsforge/vtm/blob/master/vtm-gdx/src/org/oscim/gdx/InputHandler.java#L280).

From the sample I see that there's a chance that the layer could be instance of a `GroupLayer` - how come?
We add a single `BuildingLayer` while constructing the VTM fragment - could this transform into a group layer later on, or will there be additional layers added later on, that are related to representing the 3D buildings?

**Comment by devemux86** [comment]
The sample is just an example.
The layers are created by the apps not the library.

**Comment by andrixnet** [comment]
Please include option to enable/disable 3D look for buildings.

**Comment by eddiemuc** [comment]
Hi @andrixnet ,

such a feature is already included. You find it in the map menu under "theme options:
![Screenshot_2024-09-19-06-34-27-33_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/user-attachments/assets/baa0b67d-3f9d-43d0-a527-918656a1acbb)

For Google Maps:
![Screenshot_2024-09-19-06-35-39-82_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/user-attachments/assets/2db90df1-c4e1-4668-8520-5f6b3508bd0b)

For OSM maps:
![Screenshot_2024-09-19-06-35-08-25_a916ceb0c8a82619faeb76fe6b8ae997.jpg](https://github.com/user-attachments/assets/9d549bc5-762c-4258-a360-32fcc0f7627d)

Is this what you're looking for?

**Comment by MagpieFourtyTwo** [comment]
Yep, the option is there, but it's still not working with map ``Combined Offline`` (see comment [2215410927](https://github.com/cgeo/cgeo/issues/15723#issuecomment-2215410927) which is still valid).

**Comment by SammysHP** [comment]
That's why this issue is still open. :blush:

**Comment by MagpieFourtyTwo** [comment]
I know - I just wanted to warn @andrixnet to not let expectations skyrocket too quickly.  ;)

**Comment by andrixnet** [comment]
@eddiemuc when I first reported I saw nothing in c:geo UI like that. 

Then today I updated 2024.09.18. Still no UI like that. 

Then I got digging further. (a lot of digging required, too much, in fact.)

Regarding the screenshots you've shown me: I have nothing of the kind. 
Using "default" map theme, I only get to choose between theme variants such as "default", "osmarender" (this is checked on my device, note that I never changed such setting before), etc. 
Note: "mapzen" and "openmaptiles" variants yields empty map. 

Now separate from that, I have my country's map second version from OpenAndroMaps and "Elevate.zip" theme. 
This is non-standard, it has been installed manually by me. 
Using this map, selecting the theme Elevate.zip/Elements or Elevate.zip/Elevate shows me a different visual style and under "Theme options" there is, in the bottom position "Show buildings in 3D". 

So, using "default" ... where/how can I change 3D buildings?

**Comment by eddiemuc** [comment]
> So, using "default" ... where/how can I change 3D buildings?

@andrixnet you are right. Using the default theme the option is not available. I would consider this a bug.

**Comment by moving-bits** [comment]
@andrixnet 
As a temporary workaround you can go to settings => system => view settings, enter edit mode, create a new setting named "buildingLayer3D" (watch for capitalization) of type "boolean", and set this to "false" if you want to get rid of 3D buildings.

**Comment by moving-bits** [comment]
NB: With the future of VTM being unsure, UnifiedMap probably switching to new version of Mapsforge (which has no 3D view), I would not invest in fixing this issue currently.
